core/bindings.rs

```diff
@@ -757,6 +783,52 @@ pub fn module_resolve_callback<'s>(
   None
 }
 
+// Sets the V8 promise hook, allowing tracking of context across
+// Promise chains and async functions
+fn set_promise_hook(
+  scope: &mut v8::HandleScope,
+  args: v8::FunctionCallbackArguments,
+  _rv: v8::ReturnValue,
+) {
+  let hook = match v8::Local::<v8::Function>::try_from(args.get(0)) {
+    Ok(val) => val,
+    Err(_) => {
+      let msg = v8::String::new(scope, "Invalid argument").unwrap();
+      let exception = v8::Exception::type_error(scope, msg);
+      scope.throw_exception(exception);
+      return;
+    }
+  };
+
+  let state_rc = JsRuntime::state(scope);
+  let mut state = state_rc.borrow_mut();
+
+  state.js_promise_hook_cb.push(v8::Global::new(scope, hook));
+  scope.set_promise_hook(wrapped_hook);
+}
+extern "C" fn wrapped_hook(
+  type_: v8::PromiseHookType,
+  promise: v8::Local<v8::Promise>,
+  parent: v8::Local<v8::Value>,
+) {
+  let scope = &mut unsafe { v8::CallbackScope::new(promise) };
+  let context = promise.creation_context(scope);
+  let scope = &mut v8::ContextScope::new(scope, context);
+  let global = context.global(scope);
+
+  let state_rc = JsRuntime::state(scope);
+
+  let hook_type: v8::Local<v8::Value> =
+    v8::Integer::new(scope, type_ as i32).into();
+
+  for js_promise_hook_cb_handle in state_rc.borrow().js_promise_hook_cb.iter() {
+    js_promise_hook_cb_handle
+      .get(scope)
+      .call(scope, global.into(), &[hook_type, promise.into(), parent])
```

Comment:
Maybe use `v8::undefined().into()` instead of `global.into()`? Passing `global` is not wrong and it seems to be the predominant pattern in deno_core but it's marginally less efficient.